### PR TITLE
Feat curb 3804 implement cart api

### PR DESCRIPTION
**Pull request by Rekhyt**
# Pull Request Template

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- [ ] Bug fix
- [x] New feature
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I updated the CHANGELOG.md

### CODE REVIEW

File: extension/package.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
   "dependencies": {
     "json-bigint-native": "^1.2.0",
     "request": "^2.88.2",
```
[the remark this answers is not in the document]
Comment (Rekhyt) [reply extension/package.json:20]: Stuck on old NodeJS I fear we can't get rid of `request`, `npm` won't install peer dependencies, just warn about them.

For the dev dependencies, absolutely, will look into it.

File: extension/cart/getCheckoutUrl.js [hunk 2]
```diff
@@ -1,27 +1,13 @@
...
-  // We need to replace the web_url from data.checkout with the shopify shop domain (using the alias from our config)
-  const checkoutDomain = input.checkout.web_url.replace(context.config.shopifyShopAlias + '.myshopify.com', ConfigHelper.getHostName(context.config))
+  // Once we left for the checkout there's no way of knowing whether the checkout has been completed (invalidating the
+  // cart ID) or aborted (keeping the cart ID valid). Setting a flag that will cause the initShopifyCart step to fetch
+  // the cart from Shopify Cart API to see if it is still valid.
+  const storage = context.meta.userId ? context.storage.user : context.storage.device
+  await storage.set('cartMayBeInvalid', true)
 
-  return new Url(checkoutDomain, expires)
+  return { url: `${input.shopifyCart.checkoutUrl}&logged_in=true` }
```
[the remark this answers is not in the document]
Comment (Rekhyt) [reply extension/cart/getCheckoutUrl.js:12]: You're right, Shopify might change that, they specifically advise not to rely on the link structure. I'll add parsing and rebuilding the URL to be sure.

File: extension/cart/migrateCartContents.js [hunk 3]
```diff
@@ -1,122 +1,49 @@
...
-    if (err && Object.hasOwnProperty.call(err, 'code') && err.code !== 422) {
-      context.log.error(
-        'Couldn\'t update checkout with id ' + targetCartId + ' failed with error: ' + JSON.stringify(err)
-      )
+  if (deviceCart === null) {
+    context.log.warn('Shopify API returned null when getting the device cart, creating a new one')
+    try {
+      deviceCartId = await storefrontApi.createCart()
+      deviceCart = await storefrontApi.getCart(deviceCartId)
+      await context.storage.device.set('shopifyCartId', deviceCartId)
```
[the remark this answers is not in the document]
Comment (Rekhyt) [reply extension/cart/migrateCartContents.js:26]: True, will fix

File: extension/lib/ShopifyStorefrontApi.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    if (userErrors.length === 0) return
+
+    const error = new CartError()
+    for (const userError of userErrors) {
+      const fieldIndex = (userError.field || {})[1]
+      const entityId = fieldIndex || fieldIndex === 0
+        ? (referencedContents[userError.field[1]] || {}).merchandiseId
+        : undefined
+
+      error.errors.push({ message: userError.message, entityId })
```
[the remark this answers is not in the document]
Comment (Rekhyt) [reply extension/lib/ShopifyStorefrontApi.js:149]: Done

File: extension/lib/ShopifyAdminApi.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
+const jsonb = require('json-bigint-native')
```
[the remark this answers is not in the document]
Comment (Rekhyt) [reply extension/lib/ShopifyAdminApi.js:1]: Done

File: extension/lib/ShopifyApiTokenManager.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    return token
+  }
+
+  /**
+   * Saves the COMMON Storefront API access token for app access to the extension storage
+   *
+   * @param {string} storefrontAccessToken
+   * @returns {Promise<void>}
+   */
+  async setStorefrontApiAccessToken (storefrontAccessToken) {
```
[the remark this answers is not in the document]
Comment (Rekhyt) [reply extension/lib/ShopifyApiTokenManager.js:39]: Removed it

File: extension/lib/ShopifyAdminApi.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+class ShopifyAdminApi {
+  /**
+   * @param {string} shopUrl
+   * @param {string} accessToken
+   * @param {SDKContextLog} logger
+   * @param {string?} apiVersion
+   */
+  constructor (shopUrl, accessToken, logger, apiVersion = '2024-07') {
+    this.apiUrl = `${shopUrl.replace(/\/+$/, '')}/admin/api/${apiVersion}/`
```
[the remark this answers is not in the document]
Comment (Rekhyt) [reply extension/lib/ShopifyAdminApi.js]: Done

File: extension/lib/ShopifyStorefrontApi.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+class ShopifyStorefrontApi {
+  /**
+   * @param {string} shopUrl
+   * @param {ShopifyApiTokenManager} tokenManager
+   * @param {SDKContextLog} logger A generic logger instance, e.g. current step context's .log property.
+   * @param {string?} apiVersion
+   */
+  constructor(shopUrl, tokenManager, logger, apiVersion = '2024-07') {
+    this.apiUrl = `${shopUrl.replace(/\/+$/, '')}/api/${apiVersion}/graphql.json`
```
[the remark this answers is not in the document]
Comment (Rekhyt) [reply extension/lib/ShopifyStorefrontApi.js]: Done

**Review by msandrock — APPROVED** [review]

**Pull request merged**